File: DCP/DCP/Day 5/Day5.py
#!/usr/bin/env python3

def cons(a, b):
    def pair(f):
        return f(a, b)
    return pair

def car(pair):
    return pair(lambda a, b: a)

def cdr(pair):
    return pair(lambda a, b: b)

pair = cons(3, 4)
print(car(pair) == 3)
print(cdr(pair) == 4)

File: DCP/DCP/Day 5/Day5.kt
fun main() {
    val pair = cons(3, 4)
    println(car(pair) == 3)
    println(cdr(pair) == 4)
}

fun <A, B> cons(a: A, b: B): () -> Pair<A, B> {
    return { Pair(a, b) }
}

fun <A, B> car(pair: () -> Pair<A, B>): A {
    val (a, _) = pair()
    return a
}

fun <A, B> cdr(pair: () -> Pair<A, B>): B {
    val (_, b) = pair()
    return b
}